### PR TITLE
[PIM-6420] Fix autosorting of attribute options

### PR DESCRIPTION
**Pull request by pierallard**
If fixes the autosort at 2 places:
- on attribute edition, you can enable/disable autosort, and it displays result in the list directly. Old order are kept, so you can disable autosort to get back to old one.
- on PEF, attributes are displayed ordered when option is enabled.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added Behats                      | y
| Added integration tests           | -
| Changelog updated                 | y
| Review and 2 GTM                  | Todo
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js [hunk 1]
```diff
@@ -287,7 +290,11 @@ define(
                     code_label: __('Code')
                 }));
 
-                _.each(this.collection.models, function (attributeOptionItem) {
+
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js]: extra lines to remove

File: features/Context/Page/Base/Form.php [hunk 2]
```diff
@@ -494,25 +494,41 @@ public function fillPopinFields($fields)
...
+            if (null === $container) {
+                return false;
+            }
+
+            return $container->find('css', '.select2, .select2-default');
         }, 'Impossible to find the select');
         $select2 = $this->decorate($select2, [Select2Decorator::class]);
         $selectChoices = $select2->getAvailableValues();
+
         if ($isExpected) {
```
Comment (SamirBoulil) [inline features/Context/Page/Base/Form.php]: ifififi

File: features/attribute/option/sort_attribute_options.feature [hunk 3]
```diff
@@ -13,18 +13,39 @@ Feature: Sort attribute options
...
+    And I save the attribute
+    And I should not see the text "There are unsaved changes"
+    And I am on the products page
+    And I create a new product
+    And I fill in the following information in the popin:
+      | SKU | a_product |
+    And I press the "Save" button in the popin
+    And I should be on the product "a_product" edit page
+    And I add available attributes size
+    And I should see the ordered choices [large_size], [medium_size], [small_size] in size
```
Comment (SamirBoulil) [inline features/attribute/option/sort_attribute_options.feature]: Then

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js (hunk 1 above)
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js]: extra line

File: features/Context/Page/Base/Form.php [hunk 4]
```diff
@@ -494,25 +494,41 @@ public function fillPopinFields($fields)
...
+            if ($strict) {
+                if ($selectChoices !== $choices) {
```
Comment (anaelChardan) [inline features/Context/Page/Base/Form.php]: `if ($isExpected && $strict && $selectChoices !== $choices)`

File: features/attribute/option/sort_attribute_options.feature [hunk 5]
```diff
@@ -13,18 +13,39 @@ Feature: Sort attribute options
...
+      | medium_size |
+      | large_size  |
+    And I save the attribute
+    And I should not see the text "There are unsaved changes"
+    And I am on the products page
+    And I create a new product
+    And I fill in the following information in the popin:
+      | SKU | a_product |
+    And I press the "Save" button in the popin
+    And I should be on the product "a_product" edit page
```
Comment (anaelChardan) [inline features/attribute/option/sort_attribute_options.feature]: `Then` and on the next line `when`

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js [hunk 6]
```diff
@@ -287,7 +290,11 @@ define(
...
+
+                _.each(_.sortBy(this.collection.models, function (attributeOptionItem) {
+                    return !this.sortable ? attributeOptionItem.attributes.code : 0;
```
Comment (anaelChardan) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js]: I would prefer a "truethy" condition, `this.sortable ? 0 : attributeOptionItem.attributes.code` :)

File: features/Context/DataGridContext.php [hunk 7]
```diff
@@ -769,8 +769,10 @@ public function iShouldSeeOptionInFilter($optionNames, $filterName)
 
     /**
      * @param boolean $not
-     * @param string  $option
-     * @param string  $filterName
+     * @param string $option
+     * @param string $filterName
+     *
+     * @throws ExpectationException
```
Comment (anaelChardan) [inline features/Context/DataGridContext.php]: :+1:

File: features/attribute/option/sort_attribute_options.feature [hunk 8]
```diff
@@ -13,18 +13,39 @@ Feature: Sort attribute options
       | Code            | size  |
       | Attribute group | Other |
     And I visit the "Values" tab
-    Then I should see the "Options" section
-    Then I should see "To manage options, please save the attribute first"
+    And I should see the "Options" section
```
Comment (pierallard) [inline features/attribute/option/sort_attribute_options.feature]: I never put "When/Then" on Background ; the goal of a background is not to check things, it's only to set things. So "Given" and "And" are OK IMHO.

File: features/attribute/option/sort_attribute_options.feature [hunk 9]
```diff
@@ -13,18 +13,39 @@ Feature: Sort attribute options
...
+
+  Scenario: Display attribute options ordered in PEF
+    Given I check the "Automatic option sorting" switch
+    When I create the following attribute options:
+      | Code        |
+      | small_size  |
+      | medium_size |
+      | large_size  |
+    And I save the attribute
+    And I should not see the text "There are unsaved changes"
```
Comment (pierallard) [inline features/attribute/option/sort_attribute_options.feature]: I only set Then when I have to check a result. Here, I don't care about this line, it's just a scenario line.

File: features/attribute/option/sort_attribute_options.feature (hunk 5 above)
Comment (pierallard) [inline features/attribute/option/sort_attribute_options.feature]: Same comment than above.

File: features/attribute/option/sort_attribute_options.feature (hunk 9 above)
Comment (anaelChardan) [inline features/attribute/option/sort_attribute_options.feature]: "I should" aims to check a result :)

File: features/attribute/option/sort_attribute_options.feature (hunk 5 above)
Comment (anaelChardan) [inline features/attribute/option/sort_attribute_options.feature]: Same response than above

**Pull request merged**